src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Services/DataSetVersionMappingService.cs

```diff
@@ -0,0 +1,341 @@
+
+        var sourceLocationMeta =
+            await GetLocationMeta(liveVersion.Id, cancellationToken);
+
+        var locationMappings = CreateLocationMappings(
+            sourceLocationMeta,
+            nextVersionMeta.Locations);
+
+        var sourceFilterMeta =
+            await GetFilterMeta(liveVersion.Id, cancellationToken);
+
+        var filterMappings = CreateFilterMappings(
+            sourceFilterMeta,
+            nextVersionMeta.Filters);
+
+        nextVersion.MetaSummary = nextVersionMeta.MetaSummary;
+
+        publicDataDbContext
+            .DataSetVersionMappings
+            .Add(new DataSetVersionMapping
+            {
+                SourceDataSetVersionId = liveVersion.Id,
+                TargetDataSetVersionId = nextDataSetVersionId,
+                LocationMappingPlan = locationMappings,
+                FilterMappingPlan = filterMappings
+            });
+
+        await publicDataDbContext.SaveChangesAsync(cancellationToken);
+        return Unit.Instance;
+    }
+
+    public async Task ApplyAutoMappings(
+        Guid nextDataSetVersionId,
+        CancellationToken cancellationToken = default)
+    {
+        var mappings = await publicDataDbContext
+            .DataSetVersionMappings
+            .SingleAsync(
+                mapping => mapping.TargetDataSetVersionId == nextDataSetVersionId,
+                cancellationToken);
+
+        AutoMapLocations(mappings.LocationMappingPlan);
+        AutoMapFilters(mappings.FilterMappingPlan);
+
+        mappings.LocationMappingsComplete = !mappings
+            .LocationMappingPlan
+            .Levels
+            .Any(level => level
+                .Mappings
+                .Any(optionMapping => IncompleteMappingTypes.Contains(optionMapping.Type)));
+
+        mappings.FilterMappingsComplete = !mappings
+            .FilterMappingPlan
+            .Mappings
+            .Any(filterMapping =>
+                filterMapping.Type == MappingType.None
+                || filterMapping
+                    .OptionMappings
+                    .Any(optionMapping => IncompleteMappingTypes.Contains(optionMapping.Type)));
+
+        await publicDataDbContext.SaveChangesAsync(cancellationToken);
+    }
+
+    private static void AutoMapLocations(LocationMappingPlan locationPlan)
+    {
+        locationPlan
+            .Levels
+            .ForEach(level => level
+                .Mappings
+                .ForEach(locationMapping => AutoMapElement(
+                    mapping: locationMapping,
+                    candidates: level.Candidates)));
+    }
+
+    private static void AutoMapFilters(FilterMappingPlan filtersPlan)
+    {
+        filtersPlan
+            .Mappings
+            .ForEach(filterMapping => AutoMapParentAndOptions(
+                parentMapping: filterMapping,
+                parentCandidates: filtersPlan.Candidates,
+                candidateOptionsSupplier: autoMappedCandidate => autoMappedCandidate.Options));
```

Comment:
Yeah it's useful for FilterOptions because in order to work out which candidates are valid, you firstly have to see which *Filter* candidate is mapped to the FilterOption's owning Filter, and then draw the available FilterOption candidates from that.  It's simpler with LocationOptions, because both Mappings and Candidates are grouped under their GeoLevels, so we already know which candidates should be available immediately.

Reply:
makes sense!